nifi-nar-bundles/nifi-box-bundle/nifi-box-processors/src/test/java/org/apache/nifi/processors/box/ListBoxFilesIT.java

```diff
@@ -0,0 +1,164 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.processors.box;
+
+import com.box.sdk.BoxFolder;
+import org.apache.nifi.util.MockFlowFile;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+/**
+ * See Javadoc {@link AbstractBoxFilesIT} for instructions how to run this test.
+ */
+public class ListBoxFilesIT extends AbstractBoxFilesIT<ListBoxFiles> {
+    @BeforeEach
+    public void init() throws Exception {
+        super.init();
+        testRunner.setProperty(ListBoxFiles.FOLDER_ID, mainFolderId);
+        testRunner.setProperty(ListBoxFiles.USER_ID, USER_ID);
+        testRunner.setProperty(ListBoxFiles.BOX_CONFIG_FILE, BOX_CONFIG_FILE);
+    }
+
+    @Override
+    public ListBoxFiles createTestSubject() {
+        ListBoxFiles testSubject = new ListBoxFiles();
+
+        return testSubject;
+    }
+
+    @Test
+    void listFilesFrom3LayerDeepDirectoryTree() throws Exception {
+        // GIVEN
+        BoxFolder.Info main_sub1 = createFolder("main_sub1", mainFolderId);
+        BoxFolder.Info main_sub2 = createFolder("main_sub2", mainFolderId);
+
+        BoxFolder.Info main_sub1_sub1 = createFolder("main_sub1_sub1", main_sub1.getID());
+
+        createFileWithDefaultContent("main_file1", mainFolderId);
+        createFileWithDefaultContent("main_file2", mainFolderId);
+        createFileWithDefaultContent("main_file3", mainFolderId);
+
+        createFileWithDefaultContent("main_sub1_file1", main_sub1.getID());
+
+        createFileWithDefaultContent("main_sub2_file1", main_sub2.getID());
+        createFileWithDefaultContent("main_sub2_file2", main_sub2.getID());
+
+        createFileWithDefaultContent("main_sub1_sub1_file1", main_sub1_sub1.getID());
+        createFileWithDefaultContent("main_sub1_sub1_file2", main_sub1_sub1.getID());
+        createFileWithDefaultContent("main_sub1_sub1_file3", main_sub1_sub1.getID());
+
+        Set<String> expectedFileNames = new HashSet<>(Arrays.asList(
+            "main_file1", "main_file2", "main_file3",
+            "main_sub1_file1",
+            "main_sub2_file1", "main_sub2_file2",
+            "main_sub1_sub1_file1", "main_sub1_sub1_file2", "main_sub1_sub1_file3"
+        ));
+
+        // The creation of the files are not (completely) synchronized.
+        Thread.sleep(2000);
+
+        // WHEN
+        testRunner.run();
+
+        // THEN
+        List<MockFlowFile> successFlowFiles = testRunner.getFlowFilesForRelationship(ListBoxFiles.REL_SUCCESS);
+
+        Set<String> actualFileNames = successFlowFiles.stream()
```

Comment:
Please consider extracting this call to a method since the same logic is used to collect actual filenames in 4 places.